### PR TITLE
Outliers

### PR DESCRIPTION
**Pull request by nataly-r**
- Adds a selector for outliers with on/off options
- Updates the color scale, removing grey so that it can be used to color outliers
- For k-means, the cluster means are used to determine the distance
- For dbscan, elements labeled as outliers are assigned to the closest cluster based on cluster medians

### CODE REVIEW

File: clustering.py [hunk 1]
```diff
@@ -170,13 +173,17 @@ def kmeans(data):
     tuning_ratio, tuning_min_clusters = len(data) // 20, 6
     kmeans_result = KMeans(n_clusters=tuning_ratio + tuning_min_clusters,
                            random_state=0).fit(data)
-    return kmeans_result.labels_
+    labels = np.copy(kmeans_result.labels_) + 1
+    if outlier == "on":
```
Comment (iangmhill) [inline clustering.py:177]: I see that you're using a string for the outlier variable to make it easy to map from URL parameters, but it would be clearer for an argument like this to be a boolean instead.

File: main.py [hunk 2]
```diff
@@ -49,12 +50,13 @@ def cluster(algorithm, similarity, label_encoding, chart_id):
         return data
     time_series_data, label_dict, ts_to_labels = clustering.time_series_array(
         data)
-    time_series_data = clustering.preprocess(time_series_data, label_encoding,
+    time_series_data = clustering.preprocess(time_series_data, encoding,
                                              similarity, ts_to_labels)
     if algorithm == "k-means":
-        labels = clustering.kmeans(time_series_data)
+        labels = clustering.kmeans(time_series_data, outlier)
```
Comment (iangmhill) [inline main.py:56]: Convert the "on"/"off" URL paremeter string to a boolean True or False before passing it into the kmeans function.

File: clustering.py [hunk 3]
```diff
@@ -185,8 +192,43 @@ def dbscan(data):
...
     min_max_scaler = MinMaxScaler()
     min_max_scaled = min_max_scaler.fit_transform(data)
-    dbscan_result = DBSCAN(eps=1.2, min_samples=1).fit(min_max_scaled)
-    return dbscan_result.labels_
+    dbscan_result = DBSCAN(eps=1.2, min_samples=2).fit(min_max_scaled)
+    cluster_assignment = np.copy(dbscan_result.labels_)
+    medians = cluster_medians(data, cluster_assignment)
+
+    outlier_indexes = np.where(cluster_assignment == -1)[0]
+    cluster_assignment += 1
```
Comment (iangmhill) [inline clustering.py:200]: Sorry if this is my own ignorance, but what does adding 1 to an array do in Numpy?

**Review by iangmhill — APPROVED** [review]

File: clustering.py (hunk 3 above)
Comment (nataly-r) [reply clustering.py:200]: It adds one to each element in the array. I do this so that time series can be labeled as "-clusterLabel", which would not work with the origin labels because of -0.
